File: app/src/main/java/com/mansuiki/pybot/Communicator.java
package com.mansuiki.pybot;

import android.util.Log;

import com.mansuiki.pybot.BotManager.KakaoData;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.Body;
import retrofit2.http.POST;

public class Communicator implements Observer {
    private static final String TAG = "Communicator";
    private static final String BASE_URL = "http://mansuiki.com:37280";
    private static Retrofit retrofit = null;
    private static RetrofitInterface service = null;
    private boolean isRun = false;


    Communicator() {
        retrofit = new Retrofit.Builder().baseUrl(BASE_URL).addConverterFactory(GsonConverterFactory.create()).build();
        service = retrofit.create(RetrofitInterface.class);
    }

    void start() {
        BotManager.getManager().addObserver(this);
        BotManager.getManager().resetData();
        isRun = true;
    }

    void stop() {
        BotManager.getManager().delObserver(this);
        isRun = false;
    }

    private void Upload(final KakaoData reqModel) {
        service.upload(reqModel).enqueue(new Callback<ResultModel>() {
            @Override
            public void onResponse(Call<ResultModel> call, Response<ResultModel> response) {
                ResultModel result = response.body();
                if (result.getResult().equals("BOTONPYBOT")) {
                    isRun = true;
                    KakaoListener.send(reqModel.room, "BOTON");
                } else if (result.getResult().equals("BOTOFFPYBOT")) {
                    isRun = false;
                    KakaoListener.send(reqModel.room, "BOTOFF");
                } else if (isRun) {
                    KakaoListener.send(reqModel.room, result.getResult());
                }
                Log.d(TAG, "Get Request");
            }

            @Override
            public void onFailure(Call<ResultModel> call, Throwable t) {
                Log.e(TAG, "onFailure: Failed....", t);
            }
        });
    }

    @Override
    public void update() {
        for (KakaoData data : BotManager.getManager().getSavedMsg().toArray(new KakaoData[0])) {
            BotManager.getManager().removeSavedMsg(0, data);
            if (data.message.charAt(0) == '/')
                Upload(data);
        }
    }


    interface RetrofitInterface {

        @POST("/bot")
        Call<ResultModel> upload(@Body KakaoData kakaoData);
    }

    class ResultModel {

        String Result;

        ResultModel(String Result) {
            this.Result = Result;
        }

        String getResult() {

            return Result;
        }

    }
}

File: app/src/main/java/com/mansuiki/pybot/ui/background/Communicator.kt
package com.mansuiki.pybot.ui.background

import android.util.Log
import com.mansuiki.pybot.api.ApiManager
import com.mansuiki.pybot.entity.retrofit.RequestModel
import java.util.*

class Communicator {
    companion object {
        const val TAG = "Communicator"
        private val api = ApiManager.api
    }

    private var thread: CThread? = null

    fun start() {
        thread = CThread()
        thread!!.start()
    }

    fun stop() {
        thread!!.stopc()
        thread = null
    }

    private fun Upload(reqModel: RequestModel) {
        val callSync = api.upload(reqModel)
        try {
            val response = callSync!!.execute()
            val result = response.body()
            KakaoListener.send(reqModel.room, result!!.result)
            Log.d(TAG, "Get Request")
        } catch (e: Exception) {
            Log.e(TAG, "onFailure: Failed....")
        }
        //
        /*
        service.upload(reqModel).enqueue(new Callback<ResultModel>() {
            @Override
            public void onResponse(Call<ResultModel> call, Response<ResultModel> response) {
                ResultModel result = response.body();
                KakaoListener.send(reqModel.getRoom(), result.getResult());
                Log.d(TAG, "Get Request");
            }

            @Override
            public void onFailure(Call<ResultModel> call, Throwable t) {
                Log.e(TAG, "onFailure: Failed....");
            }
        });*/
    }

    private inner class CThread : Thread() {
        private var isRun = true
        private var manager: BotManager? = null
        fun stopc() {
            isRun = false
            manager = null
            Log.d(TAG, "CThread: Stop Thread")
        }

        override fun run() {
            isRun = true
            Log.d(TAG, "CThread : Start!")
            manager!!.resetData()
            try {
                while (isRun) {
                    val temp = transData()
                    for (request in temp.toTypedArray()) {
                        Upload(request)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                Log.e(TAG, "run: ", e)
            }
        }

        private fun transData(): ArrayList<RequestModel> {
            val temp = BotManager.savedMsg
            val result = ArrayList<RequestModel>()
            var RM: RequestModel
            for (data in temp.toTypedArray()) {
                RM = RequestModel(data.room, data.sender, data.message)
                result.add(RM)
                manager!!.removeSavedMsg(0, data)
            }
            return result
        }

        init {
            manager = BotManager.manager
        }
    }
}